### PR TITLE
Enable distibuted computations for LogisticRegression

### PR DESCRIPTION
**Pull request by avolkov-intel**
# Description
Update objective_function primitive and logistic_regression algorithm to enable distributed computations. Also add spmd tests for both primitve and algorithm

### CODE REVIEW

File: cpp/oneapi/dal/algo/logistic_regression/test/spmd_fixture.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        const train_input_t input{ std::forward<Args>(args)... };
+
+        const auto split_x =
+            te::split_table_by_rows<float_t>(this->get_policy(), input.get_data(), split_count);
+        const auto split_y = te::split_table_by_rows<float_t>(this->get_policy(),
+                                                              input.get_responses(),
+                                                              split_count);
+
+        std::vector<train_input_t> split_input;
+        split_input.reserve(split_count);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/logistic_regression/test/spmd_fixture.hpp:52]: Microoptimizations in tests) (Looks fine to me though)

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 2]
```diff
@@ -610,6 +639,15 @@ sycl::event logloss_hessian_product<Float>::compute_without_fit_intercept(
         last_iter_deps = { update_grad_e };
     }
 
+    if (comm_.get_rank_count() > 1) {
+        sycl::event::wait_and_throw(last_iter_deps);
+        {
+            ONEDAL_PROFILER_TASK(hessp_allreduce);
+            auto hessp_arr = dal::array<Float>::wrap(q_, out.get_mutable_data(), out.get_count());
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:646]: I believe you [can pass event list](https://github.com/oneapi-src/oneDAL/blob/14ad1ababe5601a62983c49b0ac2e046c990c445/cpp/oneapi/dal/array.hpp#L201) as an arument to the wrap function.

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 3]
```diff
@@ -657,6 +695,33 @@ logloss_function<Float>::logloss_function(sycl::queue q,
...
+          labels_(labels),
+          n_(data.get_row_count()),
+          p_(data.get_column_count()),
+          L2_(L2),
+          fit_intercept_(fit_intercept),
+          bsz_(bsz == -1 ? get_block_size(n_, p_) : bsz),
+          hessp_(q, comm, data, L2, fit_intercept, bsz_),
+          dimension_(fit_intercept ? p_ + 1 : p_) {
+    // std::cout << "Rank" << std::endl;
+    // std::cout << comm_.get_rank() << std::endl;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:718]: Need to be removed

File: cpp/oneapi/dal/backend/primitives/objective_function/test/spmd_fixture.hpp [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        const std::int64_t block_size_tail = row_count % split_count;
+
+        std::vector<std::pair<table, ndview<std::int32_t, 1>>> result(split_count);
+
+        std::int64_t row_offset = 0;
+        for (std::int64_t i = 0; i < split_count; i++) {
+            const std::int64_t tail = std::int64_t(i + 1 == split_count) * block_size_tail;
+            const std::int64_t block_size = block_size_regular + tail;
+            // std::cout << i << ": " << row_offset << " " << block_size << std::endl;
+            REQUIRE(block_size > 0);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/spmd_fixture.hpp:47]: Please use natural order wherever it is possible:
```suggestion
            REQUIRE(0l < block_size);
```

File: samples/oneapi/dpc/ccl/data/df_binary_classification_test_data.csv
Comment (KulikovNikita) [inline samples/oneapi/dpc/ccl/data/df_binary_classification_test_data.csv:1]: Is this dataset somewhat ordered

File: samples/oneapi/dpc/ccl/sources/logistic_regression_distr_ccl.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+namespace dal = oneapi::dal;
+namespace lr = dal::logistic_regression;
+namespace result_options = dal::logistic_regression::result_options;
+
+void run(sycl::queue &queue) {
+    const auto train_data_file_name = get_data_path("df_binary_classification_train_data.csv");
+    const auto train_response_file_name = get_data_path("df_binary_classification_train_label.csv");
+    const auto test_data_file_name = get_data_path("df_binary_classification_test_data.csv");
+    const auto test_response_file_name = get_data_path("df_binary_classification_test_label.csv");
```
Comment (KulikovNikita) [inline samples/oneapi/dpc/ccl/sources/logistic_regression_distr_ccl.cpp:39]: You make the same job for all the devices

File: cpp/oneapi/dal/algo/logistic_regression/test/spmd_dpc.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+#include "oneapi/dal/algo/logistic_regression/test/spmd_fixture.hpp"
+
+namespace oneapi::dal::logistic_regression::test {
+
+TEMPLATE_LIST_TEST_M(log_reg_spmd_test, "LogReg common flow", "[lr][spmd]", log_reg_types) {
+    SKIP_IF(this->get_policy().is_cpu());
+    SKIP_IF(this->not_float64_friendly());
+
+    this->gen_input(true, 0.5);
+    this->set_rank_count(2);
```
Comment (Alexandr-Solovev) [inline cpp/oneapi/dal/algo/logistic_regression/test/spmd_dpc.cpp:26]: Its better to increase test coverage with adding     this->set_rank_count(GENERATE(2, 4));

**Review by Alexandr-Solovev — APPROVED** [review]
LGTM, with minor comment related to tests. And as an option you can mention this changes in docs

**Pull request merged**